### PR TITLE
fix: Remove possible panic in metrics clean up.

### LINKED ISSUE #12489: bug: Duplicate remove in batch task metrics
### Describe the bug

_No response_

### Error message/log

```text
thread 'risingwave-main' panicked at 'It should not have duplicate task label.: Msg("missing label values [\"14c967f7-c5ea-455e-9638-892bc94618bc\", \"11\", \"4\"]")', src/batch/src/monitor/stats.rs:171:13
stack backtrace:
   0:     0x5dc6274a418c - std::backtrace_rs::backtrace::libunwind::trace::hc317256c0daecb22
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x5dc6274a418c - std::backtrace_rs::backtrace::trace_unsynchronized::h85008105ccd95ba6
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x5dc6274a418c - std::sys_common::backtrace::_print_fmt::hb6e7cbe31fecc91c
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x5dc6274a418c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h53097d3751045778
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x5dc62122ea9f - core::fmt::rt::Argument::fmt::hdbe0b5dfb2c09e86
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/fmt/rt.rs:138:9
   5:     0x5dc62122ea9f - core::fmt::write::h4ae17c68778671a0
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/fmt/mod.rs:1094:21
   6:     0x5dc62749f04f - std::io::Write::write_fmt::hfd7c55e775c6dc60
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/io/mod.rs:1713:15
   7:     0x5dc6274a3f95 - std::sys_common::backtrace::_print::h81eabb0036b692a0
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x5dc6274a3f95 - std::sys_common::backtrace::print::h1b1f4d1a23fac3fd
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x5dc6274a5507 - std::panicking::default_hook::{{closure}}::hc17d2bd8f3da442c
  10:     0x5dc6274a5314 - std::panicking::default_hook::h3a32a9336113ba96
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:288:9
  11:     0x5dc626e2833b - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h5ba81f7b166082f5
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/alloc/src/boxed.rs:1999:9
  12:     0x5dc626e2833b - risingwave_rt::panic_hook::set_panic_hook::{{closure}}::h37923353a1c6d72e
                               at /risingwave/src/utils/runtime/src/panic_hook.rs:19:9
  13:     0x5dc626e2833b - std::panicking::update_hook::{{closure}}::hb843b94e6f821fd1
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:234:47
  14:     0x5dc6274a5b22 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h4413820818220cb9
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/alloc/src/boxed.rs:1999:9
  15:     0x5dc6274a5b22 - std::panicking::rust_panic_with_hook::h90e4d15277396259
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:709:13
  16:     0x5dc6274a58c4 - std::panicking::begin_panic_handler::{{closure}}::h2b743ea215b7c408
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:597:13
  17:     0x5dc6274a4566 - std::sys_common::backtrace::__rust_end_short_backtrace::h895d390908a90650
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/sys_common/backtrace.rs:151:18
  18:     0x5dc6274a5652 - rust_begin_unwind
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:593:5
  19:     0x5dc62122b6c3 - core::panicking::panic_fmt::h1cc8517ae3119d0e
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/panicking.rs:67:14
  20:     0x5dc62122bb73 - core::result::unwrap_failed::h4c9088fd4642a5e0
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/result.rs:1651:5
  21:     0x5dc621d9cc3e - core::result::Result<T,E>::expect::h6c2571cd5bd9a791
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/result.rs:1033:23
  22:     0x5dc621d9cc3e - risingwave_batch::monitor::stats::BatchTaskMetrics::clean_metrics::hd101e1f7df99529a
                               at /risingwave/src/batch/src/monitor/stats.rs:171:13
  23:     0x5dc621d9cc3e - <risingwave_batch::monitor::stats::BatchTaskMetrics as prometheus::metrics::Collector>::collect::heacba35ddfd61423
                               at /risingwave/src/batch/src/monitor/stats.rs:198:9
  24:     0x5dc62167da12 - prometheus::registry::RegistryCore::gather::hcb813cd1f0faea5e
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/prometheus-0.13.3/src/registry.rs:123:23
  25:     0x5dc62167da12 - prometheus::registry::Registry::gather::ha5db899605bab70c
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/prometheus-0.13.3/src/registry.rs:276:9
  26:     0x5dc622850108 - risingwave_common_service::metrics_manager::MetricsManager::metrics_service::{{closure}}::h5e55103fd21ff79f
                               at /risingwave/src/common/common_service/src/metrics_manager.rs:49:18
  27:     0x5dc622850108 - <hyper::proto::h1::dispatch::Server<S,hyper::body::body::Body> as hyper::proto::h1::dispatch::Dispatch>::poll_msg::he01714f94115dbea
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.26/src/proto/h1/dispatch.rs:491:35
  28:     0x5dc622850108 - hyper::proto::h1::dispatch::Dispatcher<D,Bs,I,T>::poll_write::hcbe9b9fd7d058b94
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.26/src/proto/h1/dispatch.rs:297:43
  29:     0x5dc622850108 - hyper::proto::h1::dispatch::Dispatcher<D,Bs,I,T>::poll_loop::hbd79b75d7a5e25ff
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.26/src/proto/h1/dispatch.rs:161:21
  30:     0x5dc622850108 - hyper::proto::h1::dispatch::Dispatcher<D,Bs,I,T>::poll_inner::h7e5947ee74ac8156
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.26/src/proto/h1/dispatch.rs:137:16
  31:     0x5dc622850108 - hyper::proto::h1::dispatch::Dispatcher<D,Bs,I,T>::poll_catch::hf0ad500b03d3ecd9
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.26/src/proto/h1/dispatch.rs:120:28
  32:     0x5dc622816e9d - <hyper::proto::h1::dispatch::Dispatcher<D,Bs,I,T> as core::future::future::Future>::poll::he16e6c000a0322a5
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.26/src/proto/h1/dispatch.rs:424:9
  33:     0x5dc622816e9d - <hyper::server::conn::ProtoServer<T,B,S,E> as core::future::future::Future>::poll::hbc7a9aa8e1381f43
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.26/src/server/conn.rs:986:50
  34:     0x5dc6227cdefc - <hyper::server::conn::upgrades::UpgradeableConnection<I,S,E> as core::future::future::Future>::poll::hfd29d6276279bcb8
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.26/src/server/conn.rs:1046:30
  35:     0x5dc6227be10e - <hyper::server::server::new_svc::NewSvcTask<I,N,S,E,W> as core::future::future::Future>::poll::hccbe17157913889d
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.26/src/server/server.rs:763:36
  36:     0x5dc6227cb831 - <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll::hb6fc762f0d449e4a
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  37:     0x5dc6227cb831 - tokio::runtime::task::core::Core<T,S>::poll::{{closure}}::hf603453a0dfb0448
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/core.rs:223:17
  38:     0x5dc6227cb831 - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h1ee3f8a1efeaced6
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/loom/std/unsafe_cell.rs:14:9
  39:     0x5dc62283de2a - tokio::runtime::task::core::Core<T,S>::poll::h05cca97cbf490409
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/core.rs:212:13
  40:     0x5dc6228005db - tokio::runtime::task::harness::poll_future::{{closure}}::h00fbd97c0d74ad9d
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:476:19
  41:     0x5dc6228005db - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::ha23ab183185a8f4e
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/panic/unwind_safe.rs:271:9
  42:     0x5dc6228005db - std::panicking::try::do_call::h0084bb62f3b3b38b
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:500:40
  43:     0x5dc6228005db - std::panicking::try::hb2df1124fc90601e
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:464:19
  44:     0x5dc6228005db - std::panic::catch_unwind::hd39d41169e8e2f5a
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panic.rs:142:14
  45:     0x5dc6228005db - tokio::runtime::task::harness::poll_future::h70867eab6700c113
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:464:18
  46:     0x5dc6228005db - tokio::runtime::task::harness::Harness<T,S>::poll_inner::hf5e0057dfa21e34d
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:198:27
  47:     0x5dc6228005db - tokio::runtime::task::harness::Harness<T,S>::poll::hf3646ac5d46280ac
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:152:15
  48:     0x5dc62753fceb - tokio::runtime::task::raw::RawTask::poll::h2bb4fdbafd629891
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/raw.rs:200:18
  49:     0x5dc62753fceb - tokio::runtime::task::LocalNotified<S>::run::hf4cccdab20e2ebe2
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/mod.rs:394:9
  50:     0x5dc62753fceb - tokio::runtime::scheduler::multi_thread::worker::Context::run_task::{{closure}}::h273fa1aca0c3c00c
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:464:18
  51:     0x5dc62753fceb - tokio::runtime::coop::with_budget::h0f38d48d5480b036
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/coop.rs:107:5
  52:     0x5dc62753fceb - tokio::runtime::coop::budget::h15db80c9877d09ac
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/coop.rs:73:5
  53:     0x5dc62753fceb - tokio::runtime::scheduler::multi_thread::worker::Context::run_task::ha845c8689f6ccb29
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:463:9
  54:     0x5dc62753f3e3 - tokio::runtime::scheduler::multi_thread::worker::Context::run::h3ac9374ca6c7b2ab
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:426:24
  55:     0x5dc627533bed - tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}::h93466879b0aaeca6
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:406:17
  56:     0x5dc627533bed - tokio::macros::scoped_tls::ScopedKey<T>::set::h32afbefddf4de3ba
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/macros/scoped_tls.rs:61:9
  57:     0x5dc62753eec8 - tokio::runtime::scheduler::multi_thread::worker::run::h1c7e078ea6ac8739
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:403:5
  58:     0x5dc62752241c - tokio::runtime::scheduler::multi_thread::worker::Launch::launch::{{closure}}::h91f4c0935669f66e
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:365:45
  59:     0x5dc62752241c - <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll::hd3a1237645595b5e
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/blocking/task.rs:42:21
  60:     0x5dc62752241c - <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll::hec08e13f9e216607
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  61:     0x5dc627528c1e - tokio::runtime::task::core::Core<T,S>::poll::{{closure}}::hf95c41ce6207e587
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/core.rs:223:17
  62:     0x5dc627528c1e - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h0f47cc227372088e
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/loom/std/unsafe_cell.rs:14:9
  63:     0x5dc6275230aa - tokio::runtime::task::core::Core<T,S>::poll::h18129f26fd06c282
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/core.rs:212:13
  64:     0x5dc62750af3f - tokio::runtime::task::harness::poll_future::{{closure}}::h34c60e7df3774cfe
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:476:19
  65:     0x5dc62750af3f - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hfe4eb2b6ea95f01f
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/panic/unwind_safe.rs:271:9
  66:     0x5dc62750af3f - std::panicking::try::do_call::h87072722a160a9af
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:500:40
  67:     0x5dc62750af3f - std::panicking::try::hb28191e6867eef1d
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:464:19
  68:     0x5dc62750af3f - std::panic::catch_unwind::h299da6d5116c0ae6
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panic.rs:142:14
  69:     0x5dc62750af3f - tokio::runtime::task::harness::poll_future::h5e3da918fd6a9b1b
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:464:18
  70:     0x5dc62750af3f - tokio::runtime::task::harness::Harness<T,S>::poll_inner::h0447b267cb11c04d
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:198:27
  71:     0x5dc62750af3f - tokio::runtime::task::harness::Harness<T,S>::poll::hee03aec0809d5518
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:152:15
  72:     0x5dc627508e46 - tokio::runtime::task::raw::RawTask::poll::h2bb4fdbafd629891
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/raw.rs:200:18
  73:     0x5dc627508e46 - tokio::runtime::task::UnownedTask<S>::run::h6c8629ca4541f64f
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/mod.rs:431:9
  74:     0x5dc627508e46 - tokio::runtime::blocking::pool::Task::run::hdd4790cf20e934f8
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/blocking/pool.rs:159:9
  75:     0x5dc627508e46 - tokio::runtime::blocking::pool::Inner::run::hf61b7bfb1d50331a
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/blocking/pool.rs:513:17
  76:     0x5dc6275327d5 - tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}::he4eb394ddedef5ee
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/blocking/pool.rs:471:13
  77:     0x5dc6275327d5 - std::sys_common::backtrace::__rust_begin_short_backtrace::hd896fc9835dc356a
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/sys_common/backtrace.rs:135:18
  78:     0x5dc627516502 - std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}::h570a304190bf6be5
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/thread/mod.rs:529:17
  79:     0x5dc627516502 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h4c64fa95b7eff8c6
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/panic/unwind_safe.rs:271:9
  80:     0x5dc627516502 - std::panicking::try::do_call::hbd8a12db9d29d2ce
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:500:40
  81:     0x5dc627516502 - std::panicking::try::h04bf2a00e641d883
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:464:19
  82:     0x5dc627516502 - std::panic::catch_unwind::hbc6516b662b7143c
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panic.rs:142:14
  83:     0x5dc627516502 - std::thread::Builder::spawn_unchecked_::{{closure}}::h19598c8db1771041
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/thread/mod.rs:528:30
  84:     0x5dc627516502 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h7a4c28577f32cee1
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/ops/function.rs:250:5
  85:     0x5dc6274ab895 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h0716d70b901d1398
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/alloc/src/boxed.rs:1985:9
  86:     0x5dc6274ab895 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h7c11e50091aab2e1
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/alloc/src/boxed.rs:1985:9
  87:     0x5dc6274ab895 - std::sys::unix::thread::Thread::new::thread_start::h3ae99d1c2785d9c6
                               at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/sys/unix/thread.rs:108:17
  88:     0x70242366bb43 - start_thread
                               at ./nptl/pthread_create.c:442:8
  89:     0x7024236fcbb4 - clone
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:100
  90:                0x0 - <unknown>
```
```


### To Reproduce

_No response_

### Expected behavior

_No response_

### How did you deploy RisingWave?

_No response_

### The version of RisingWave

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by liurenjie1024**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Remove possible panic in metrics cleanup.

Close #12489  

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by wenym1 — COMMENTED** [review]
Just introduce a label guarded metrics in https://github.com/risingwavelabs/risingwave/pull/12882 that will automatically remove the label when the metrics variable is dropped. May consider porting these metrics to use the label guarded one?

**Comment by liurenjie1024** [comment]
> Just introduce a label guarded metrics in #12882 that will automatically remove the label when the metrics variable is dropped. May consider porting these metrics to use the label guarded one?

Done.

**Comment by liurenjie1024** [comment]
cc @wenym1 Any other comments?

**Review by wenym1 — COMMENTED** [review]
I just realize that the current label guarded metrics may drop the label before it is collected, which will cause a loss of metrics. I will have a new PR to fix this first.

**Comment by wenym1** [comment]
> I just realize that the current label guarded metrics may drop the label before it is collected, which will cause a loss of metrics. I will have a new PR to fix this first.

Can be fixed by https://github.com/risingwavelabs/risingwave/pull/13080

File: src/batch/src/task/context.rs [hunk 1]
```diff
@@ -177,7 +181,9 @@ impl ComputeNodeContext {
             batch_metrics
```
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply src/batch/src/task/context.rs:181]: Sorry, I don't quite get your point. Could you explain why?

[the remark this answers is not in the document]
Comment (wenym1) [reply src/batch/src/task/context.rs:181]: It's like
```
struct LabelGuardedGauge {
    gauge: Gauge,
    ...
}

fn f() -> Gauge {
  let label_guarded_gauge = LabelGuardedGauge::new();
  let gauge: Gauge = label_guarded_gauge.gauge.clone();
  return gauge;
}
```

The LabelGuardedGauge removes the label when it gets dropped. In the code it is only a local temporary variable and gets dropped immediately after the method returns.

**Comment by wenym1** [comment]
Support for label guarded Gauge is supported in https://github.com/risingwavelabs/risingwave/pull/13080 by the way. Related code in this PR can be removed.

File: src/batch/src/task/context.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (liurenjie1024) [reply src/batch/src/task/context.rs]: I see, thanks!

**Comment by liurenjie1024** [comment]
> Support for label guarded Gauge is supported in #13080 by the way. Related code in this PR can be removed.

I think it's already removed in this pr.

**Review by wenym1 — APPROVED** [review]
Rest LGTM!

**Pull request merged**